pkg/docker/docker.go

```diff
@@ -0,0 +1,35 @@
+package docker
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/giantswarm/e2e-harness/pkg/harness"
+)
+
+func Run(imageTag string, out io.Writer, entrypoint string, args ...string) error {
+	args = append([]string{"quay.io/giantswarm/e2e-harness:" + imageTag}, args...)
```

Comment:
Why not use shipyard as a library instead?

Reply:
I understand you want to use the docker image as an aggregator of binaries. I think we can do better here.
My idea is to pass path to the binaries as CLI arguments defaulting to paths present in the image. That way we don't have to pass additional imageTag and make the code less confusing.

Reply:
Sorry, I'm missing your point here. `quay.io/giantswarm/e2e-haeness` is built from this same repo, this is the Dockerfile in this PR https://github.com/giantswarm/e2e-harness/blob/946e47b607903afb304263c77eef281f77194b13/Dockerfile. It gives a common, well-known environment for all the repos using e2e-harness to run e2e-tests, atm it includes shipyard, kubectl and helm binaries, available in the default `PATH`. 

Now, the image tag is used to establish the version of the image being used, in the e2e-harness repo it will allow us to test the e2e-harness image just built by architect in the first CI step, in the clilent repos to fix the image version used to a well-known one. Why do we need to pass the path of the binaries? Why would this replace passing the imageTag?

Reply:
You download the binary for a version. Then you run it passing the same version as `image-tag` argument, because the binary runs other binaries from the image. That way you have to maintain two distribution channels, docker images, and binaries, which looks odd to me personally. Especially when I read the code. 

What I'm proposing is making this opposite. Instead of `image-tag` flag `e2e-harness` takes `shipyard-path`, `kubectl-path` and `helm-path` flags, defaulting to `/bin/shipyard`, `/bin/kubectl` and `/bin/helm` respectively. `e2e-harness` runs the binaries directly instead of invoking docker image of itself.  Then when you build the image you make sure binaries are in default paths,  you add `e2e-harness` as entrypoint and you can run `e2e-entrypoint` from a docker image. That way you avoid running docker inside the binary, you don't have two distribution channels, and the code looks more intuitive IMO.

If I'm still not clear we can sync on Monday and talk about this.

Reply:
Ok, let's talk on monday, the problem I see with your approach is that every repo would need to know how to invoke the e2e-harness container, which paths to mount and where, which environment variables to pass, how to persist state between different calls, etc. Maybe I'm missing an easy way of doing it.

The current approach works similarly to architect, usually we would download `latest` release of the e2e-harness binary, as in https://github.com/giantswarm/aws-operator/blob/master/circle.yml#L11, and use a specific tag for the container image, as in https://github.com/giantswarm/architect/blob/master/workflow/helm.go#L16, anyway let's sync about this.

Reply:
Now I understand. You also posted it here: https://github.com/giantswarm/e2e-harness/pull/3#discussion_r146103510

The reason you mentioned is good enough for me.

There is a little nice thing we can do about it though. We can get rid of `image-tag` argument by storing SHA during build time. You can create `gitCommit` variable in main.go and use that instead `image-tag` so we have that baked in into the binary. See https://github.com/giantswarm/architect/blob/f58f67262ac7828fe15350d51c118db3821a734c/workflow/golang.go#L239 for details.

WDYT?

It took me a while to wrap my head around your reasoning. Thanks for bearing with me <3

Reply:
np at all! :) it has helped me a lot writing down all this things that were just floating around. `gitCommit` you mention sounds very nice,  /me looks...

Reply:
Done the changes for using `gitCommit` and removing the flag, pls take look. As mentioned before things change a bit in later PRs (configurable values on each package, so that gitCommit can be passed to the docker value at creation time), now to keep things simple this var is injected into the `cmd` package from main.